File: pp/tools/check_catalogue/check_catalogue.py
import numpy as np
import sys
import os

# USAGE
# This python script checks the output of a peak patch run. It can be run
# at command line by running:
#     python3 check_catalogue.py <...>/<short_name>_merge.pksc.<seed>

# Raise an error if incorrect number of arguments at command line
if len(sys.argv) != 2:
    raise ValueError('You\'ve passed me the wrong number of arguments at t'
        +'he command line. Command\nline should take the form\npython3 che'
        +'ck_catalogue.py <...>/<short_name>_merge.pksc.<seed>')

# Open catalogue file
catalogue_file = str(sys.argv[1]).strip()
in_catalogue = open( catalogue_file, 'rb' )

# Read header from catalogue file and print
num_halos = np.fromfile( in_catalogue, dtype=np.int32,   count=1 )[0]
R_THmax   = np.fromfile( in_catalogue, dtype=np.float32, count=1 )[0]
z_thing   = np.fromfile( in_catalogue, dtype=np.float32, count=1 )[0]
print('Number of halos:  ',num_halos)
print('Max filter scale: ',R_THmax)
print('zthing:           ',z_thing)

# Read file size
bytes_per_halo_param = 4 # 4 for 32-bit floating-point numbers
header_length        = 3 # number of e.g. 32-bit floats or integers
cat_file_size   = os.stat(catalogue_file).st_size # bytes
num_halo_params = int(
    (cat_file_size/bytes_per_halo_param-header_length)/num_halos )
print('This catalogue appears to have {0} parameters per halo.'
    .format(num_halo_params))

# Read catalogue body
catalogue = np.fromfile( in_catalogue, dtype=np.float32,
    count=num_halo_params*num_halos )
catalogue = np.reshape( catalogue, (num_halos,num_halo_params) )

print('The mean value for initial position is ({0},{1},{2}) where {3}<x_L^i<{4}'
    .format( str(round( np.mean(catalogue[:,7])  , 2 )) ,
             str(round( np.mean(catalogue[:,8])  , 2 )) ,
             str(round( np.mean(catalogue[:,9])  , 2 )) , 
             str(round( np.min(catalogue[:,7:9]) , 2 )) ,
             str(round( np.max(catalogue[:,7:9]) , 2 )) ))

# Print some samples of catalogue data
print('x','y','z','v_x','v_y','v_z','R_TH','x_L','y_L','z_L','F_pk',sep='\t\t')
for i in range(2):
    for j in range(11):
        if j<10:
            print(catalogue[i,j],end='\t')
        else:
            print(catalogue[i,j])

print('.\n.\n.')

for i in range(num_halos-1,num_halos):
    for j in range(11):
        if j<10:
            print(catalogue[i,j],end='\t')
        else:
            print(catalogue[i,j])

File: pp/tools/check_catalogue/check_catalogue.f90
! This programme is a quick check for 

program check_catalogue
    use :: iso_fortran_env, only : file_storage_size
    implicit none
    integer(kind=4) i, j, num_halos, num_halo_params, dummy
    real(kind=4)    R_THmax,z_thing
    real(kind=4),allocatable :: posxyz(:,:)  , velxyz(:,:) , &
                                posLxyz(:,:) , rth(:)      , F(:)
    character(len=128) catalogue,output
    character(len=:),allocatable :: tab
    tab='		'

    ! Read catalogue file name from command line
    read(*,*) catalogue

    ! Read catalogue file header and print it to the screen
    open(unit=1,file=catalogue,form='unformatted',access='stream')
    read(1) num_halos, R_THmax, z_thing
    close(1)
    write(*,*) 'Number of halos:  ' , num_halos
    write(*,*) 'Max filter scale: ' , R_THmax
    write(*,*) 'zthing:           ' , z_thing

    ! Allocate halo parameters:
    allocate(posxyz(3,num_halos))  ! Eulerian position
    allocate(velxyz(3,num_halos))  ! Velocity
    allocate(posLxyz(3,num_halos)) ! Lagrangian position
    allocate(rth(num_halos))       ! Filter scale
    allocate(F(num_halos))         ! Peak overdensity

    ! Read file size in bytes of halo catalogue, divide by 4 to get the
    ! number of 32-bit values in the catalogue, subtract the 3 values in
    ! the header, divide by the number of halos to get # 32-bit floats per
    ! halo
    inquire(file=catalogue, size=num_halo_params)
    num_halo_params=(num_halo_params/4-3)/num_halos
    write(*,*) 'The number of halo parameters is ', num_halo_params

    ! Read catalogue main body
    open(unit=1,file=catalogue,form='unformatted',access='stream')
    read(1) dummy, dummy, dummy
    read(1) ( ( posxyz(j,i),  j=1,3                ),         &
              ( velxyz(j,i),  j=1,3                ), rth(i), &
              ( posLxyz(j,i), j=1,3                ), F(i),   &
              ( dummy,        j=12,num_halo_params ), i=1,num_halos )
    close(1)

    write(*,*) 'x'//tab//'y'//tab//'z'//tab//'v_x'//tab//'v_y'//tab//'v_z'&
        //tab//'R_TH'//tab//'x_L'//tab//'y_L'//tab//'z_L'//tab//'F_pk'
    
    write(*,*) ( ( posxyz(j,i),  j=1,3 ),         &
                 ( velxyz(j,i),  j=1,3 ), rth(i), &
                 ( posLxyz(j,i), j=1,3 ), F(i),   i=1,2 ) 
    write(*,*) '.'
    write(*,*) '.'
    write(*,*) '.'
    write(*,*) ( ( posxyz(j,i),  j=1,3 ),         &
                 ( velxyz(j,i),  j=1,3 ), rth(i), &
                 ( posLxyz(j,i), j=1,3 ), F(i),   i=num_halos-1,num_halos )

end program check_catalogue
